aea/cli/registry/ipfs.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Module with methods for ipfs registry."""
+
+import json
+import os
+from pathlib import Path
+from typing import List, Optional, Union
+
+import click
+
+from aea.cli.utils.constants import LOCAL_REGISTRY_PATH
+from aea.configurations.base import PublicId
+from aea.helpers.io import open_file
+
+
+def get_ipfs_hash_from_public_id(
+    item_type: str, public_id: PublicId, local_registry_path: Path
+) -> Optional[str]:
+    """Get IPFS hash from local registry."""
+
+    with open_file(str(local_registry_path), mode="r") as fp:
+        local_registry_data = json.load(fp)
+
+    if public_id.package_version.is_latest:
+        package_versions: List[PublicId] = [
+            PublicId.from_str(_public_id)
+            for _public_id in local_registry_data.get(f"{item_type}s").keys()
+            if public_id.same_prefix(PublicId.from_str(_public_id))
+        ]
+        package_versions = list(
+            reversed(sorted(package_versions, key=lambda x: x.package_version))
+        )
+        if len(package_versions) == 0:
+            return None
+        public_id, *_ = package_versions
+
+    return local_registry_data.get(f"{item_type}s").get(str(public_id), None)
+
+
+def register_item_to_local_registry(
+    item_type: str, public_id: Union[str, PublicId], package_hash: str
+) -> None:
+    """
+    Add PublicId to hash mapping in the local registry.
+
+    :param item_type: item type.
+    :param public_id: public id of package.
+    :param package_hash: hash of package.
+    """
+
+    local_registry_path = Path(LOCAL_REGISTRY_PATH)
+    if local_registry_path.is_file():
+        with open_file(local_registry_path, mode="r") as fp:
+            local_registry_data = json.load(fp)
+    else:
+        local_registry_data = {
+            "protocols": {},
+            "skills": {},
+            "connections": {},
+            "contracts": {},
+            "agents": {},
+        }
```

Comment:
There are various failure modes I can see here. E.g. if the file exists but is corrupted

Reply:
Should I add a schema validator ?